routing/control_tower.go

```diff
@@ -85,10 +141,115 @@ func (p *controlTower) Success(paymentHash lntypes.Hash,
 func (p *controlTower) Fail(paymentHash lntypes.Hash,
 	reason channeldb.FailureReason) error {
 
-	return p.db.Fail(paymentHash, reason)
+	// Lock hash to prevent the race condition described in the Success
+	// method.
+	p.hashMutex.Lock(string(paymentHash[:]))
+	defer p.hashMutex.Unlock(string(paymentHash[:]))
+
+	err := p.db.Fail(paymentHash, reason)
+	if err != nil {
+		return err
+	}
+
+	// Notify subscribers of fail event.
+	p.notifyFinalEvent(
+		paymentHash, PaymentResult{
+			Success:       false,
+			FailureReason: reason,
+		},
+	)
+
+	return nil
 }
 
 // FetchInFlightPayments returns all payments with status InFlight.
 func (p *controlTower) FetchInFlightPayments() ([]*channeldb.InFlightPayment, error) {
 	return p.db.FetchInFlightPayments()
 }
+
+// SubscribePayment subscribes to updates for the payment with the given hash.
+// It returns a boolean indicating whether the payment is still in flight and a
+// channel that provides the final outcome of the payment.
+func (p *controlTower) SubscribePayment(paymentHash lntypes.Hash) (
+	chan PaymentResult, error) {
+
+	// Create a channel with buffer size 1. For every payment there will be
+	// exactly one event sent.
+	c := make(chan PaymentResult, 1)
+
+	// Lock the hash to prevent a race condition with subscribing and a
+	// simultaneous update from router.
+	p.hashMutex.Lock(string(paymentHash[:]))
+	defer p.hashMutex.Unlock(string(paymentHash[:]))
+
+	payment, err := p.db.FetchPayment(paymentHash)
+	if err != nil {
+		return nil, err
+	}
+
+	var event PaymentResult
+
+	switch payment.Status {
+
+	// Payment is currently in flight. Register this subscriber and
+	// return without writing a result to the channel yet.
+	case channeldb.StatusInFlight:
+		p.subscribersMtx.Lock()
+		defer p.subscribersMtx.Unlock()
+
+		subscribers := p.subscribers[paymentHash]
+		subscribers = append(subscribers, c)
+		p.subscribers[paymentHash] = subscribers
+
+		return c, nil
+
+	// Payment already succeeded. It is not necessary to register as
+	// a subscriber, because we can send the result on the channel
+	// immediately.
+	case channeldb.StatusSucceeded:
+		event.Success = true
+		event.Preimage = *payment.PaymentPreimage
+		event.Route = &payment.Attempt.Route
+
+	// Payment already failed. It is not necessary to register as a
+	// subscriber, because we can send the result on the channel
+	// immediately.
+	case channeldb.StatusFailed:
+		event.Success = false
+		event.FailureReason = *payment.Failure
```

Comment:
don't we want to set the route for failures?

Reply:
Not yet. With `timeout` and `no_route` the route itself isn't relevant. I am planning to expose the route only on failures where it makes sense, like `unknown_pay_hash`.